File: boardview.cpp
/***************************************************************************
                          boardview.h  -  description
                             -------------------
    begin                : Sat May 3 2003
    copyright            : (C) 2003 by Ilmari Heikkinen
    email                : irheikki@cc.hut.fi
 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/

#include "boardview.h"
#include "piece.h"
#include "animpiece.h"
#include "gamerules.h"

#include "aiplayer.h"

#include <qwidget.h>
#include <qevent.h>
#include <qtimer.h>


////////
// PUBLIC METHODS
//

BoardView::BoardView( QCanvas *canvas, QWidget *parent, const char *name )
    : QCanvasView( canvas, parent, name ),
    rules(new GameRules()),
    players(new QValueList<QObject*>)
{
  animations=TRUE;
  animSpeed=1;
  animRandomJitter=FALSE;

  newMove();
  newBlackSingleGame();

  animType = 2;
}

BoardView::~BoardView()
{
  // tell the player objects to bugger off and die
  disconnectAllPlayers();
  delete Timer;
}


// update status of pieces on board
// animate if animations
void BoardView::updateView()
{
  if (animations) {
      animateBoard();
  } else {
      clearBoard();
  }
  canvas()->update();
}

// update board with flip animations
void BoardView::animateBoard()
{
    QCanvasItemList list = canvas()->allItems();
    QCanvasItemList::iterator it, jt;
    double i=-1;
    double delay = 0;
    for ( it = list.begin(); it != list.end(); ++it) {
      if ( (*it)->rtti() == PIECE_RTTI )
      {
        Piece* p = (Piece*) (*it);

        // the color of piece in internal othello board
        unsigned int newFrame = rules->getPiece(  p->getPosition()  );

        if ( p->getColor() != newFrame ) {
          if (newFrame == BLANK || p->getColor() == BLANK) {
            p->setColor(newFrame);
            p->setFrame(newFrame);
          } else {
            for ( jt = list.begin(); jt != list.end(); ++jt) {
                // find animation that matches the new state of the piece
                if ( (*jt)->rtti() == ANIM_RTTI ) {
                    AnimPiece* a = (AnimPiece*) (*jt);
                    // correctly colored and in the correct position
                    if ( (a->getResult() == newFrame) && (a->getPosition() == p->getPosition()) ) {
                      if (!animRandomJitter) {
                        i = animSpeed * animDelayAlgo(animType, p, &delay);
                      } else {
                        i++;
                      }
                      a->animPlay(p,i,animRandomJitter);
                      break;
                    }
                }
            }
          }
        }
      }
    }
}

double BoardView::animDelayAlgo(unsigned int algo, Piece* p, double* delay)
{
  switch (algo) {
    case 1: // logarithmic away from center
      return ( 5.5*(log10(3.5*distanceFromNew(p->getPosition()) - 3.0)) );
      break;
    case 2: // linear away from center
      return ( distanceFromNew(p->getPosition()) - 1.0);
      break;
    case 3: // logarithmic towards center
      return (4*getMaximumDistance())/( 5.5*(log10(3.5*distanceFromNew(p->getPosition()) - 3.0)) );
      break;
    case 4: // linear towards center
      return (3*getMaximumDistance())/(1 + abs( distanceFromNew(p->getPosition()))) - 1.0;
      break;
    case 5: // linear away from center, angled
      return (distanceFromNew(p->getPosition()) -1.0) + M_PI + angleFromNew(p->getPosition());
      break;
    default:
      double retval = (*delay);
      (*delay)++;
      return (retval);
  }
}

// angle from new piece
double BoardView::angleFromNew(unsigned int pos)
{
  QValueList<unsigned int>::iterator changes = rules->getChanges();
  changes++;
  unsigned int newpos = (*changes);
  TCoord p = rules->uiToTCoord(pos);
  TCoord n = rules->uiToTCoord(newpos);

  double diff_x = p.x - n.y;
  double diff_y = p.y - n.y;

  double angle = atan2(diff_x, diff_y);
  return angle;
}

// maximum distance from new piece
double BoardView::getMaximumDistance()
{
  double maxDistance=0, temp=0;
  QValueList<unsigned int>::iterator changes = rules->getChanges();
  changes++;
  changes++;
  do {
      temp=distanceFromNew((*changes));
      if (temp > maxDistance) {
              maxDistance = temp;
      }
      changes++;
  } while ((*changes) != 0 && (*changes) < 64);
  return maxDistance;
}

// distance from new piece in pieces
double BoardView::distanceFromNew(unsigned int pos)
{
  QValueList<unsigned int>::iterator changes = rules->getChanges();
  changes++;
  unsigned int newpos = (*changes);
  TCoord p = rules->uiToTCoord(pos);
  TCoord n = rules->uiToTCoord(newpos);

  // double diff  = sqrt( pow((float)p.x-n.x,2) + pow((float)p.y-n.y,2) );
  double abs_x = abs(p.x - n.x);
  double abs_y = abs(p.y - n.y);
  return (abs_x < abs_y) ? abs_y : abs_x;
}

// set piece frames and colors according to current state, no animations
void BoardView::clearBoard()
{
  QCanvasItemList list = canvas()->allItems();
  QCanvasItemList::iterator it;
  for ( it = list.begin(); it != list.end(); ++it) {
    if ( (*it)->rtti() == PIECE_RTTI ) {
      unsigned int newFrame = rules->getPiece( ((Piece*)(*it))->getPosition() );
      ((Piece*)(*it))->setColor( newFrame );
      ((Piece*)(*it))->setFrame( newFrame );
    } else if ( (*it)->rtti() == ANIM_RTTI ) {
      ((AnimPiece*)(*it))->stopPlaying();
    }
  }
}

// return true if there's space for another player, else false
bool BoardView::isSpace()
{
  if (space > 0) {
    return true;
  } else {
    return false;
  }
}
                                      
// connect newPlayer if there's space,
// listen to newPlayer's makeMove-signal
bool BoardView::connectPlayer(QObject* newPlayer)
{
  if (isSpace()) {
        space--;
        players->append(newPlayer);
        // listen to the moves the new player makes
        connect (newPlayer, SIGNAL( makeMove(unsigned int) ),
                      this, SLOT( extMove(unsigned int) ));
        // listen to the death throes of the player
        connect (newPlayer, SIGNAL( disconnectMe(QObject*) ),
                      this, SLOT( disconnectPlayer(QObject*) ));
        return true;
       
  } else {
        return false;
  }
}

// set animation variable
void BoardView::setAnimated(bool animate)
{
  animations = animate;
}
// get animation variable
bool BoardView::getAnimated()
{
  return animations;
}
// set animation speed. if new speed is negative, use random jitter
void BoardView::setAnimSpeed(double new_speed)
{
  if (new_speed >= 0  ) {
    animSpeed = new_speed;
    animRandomJitter = FALSE;
  } else {
    animSpeed = 1;
    animRandomJitter = TRUE;
  }
}
// get animation speed
double BoardView::getAnimSpeed()
{
  return animSpeed;
}


////////
// PROTECTED METHODS
//

// process mouseclicks
void BoardView::contentsMousePressEvent( QMouseEvent *e )
{
    // a list of all items underneath the mouse when it was pressed
    QCanvasItemList list = canvas()->collisions(e->pos());

    // RTTI of a QCanvasSprite is 1
    if ( !list.isEmpty() && list.first()->rtti() == PIECE_RTTI )
    {
       Piece* p = (Piece*)list.first();
       if (p->frame() == BLANK) {
         unsigned int pos = p->getPosition();
         if (player != BLANK &&
             (player == BOTH || rules->getTurn() == player) &&
             makeMove(pos)
            ) { // something to process after move made?
         }
       }
    }    
}


////////
// PROTECTED SLOTS
//

// process external move calls
void BoardView::extMove(unsigned int pos) {
  if (rules->getTurn() != player || // not local player's turn, ie. external player's turn
      player == BLANK) { // [computer vs computer] or [network client vs network client]
      makeMove(pos);
  }
}

// gametype call wrappers for the menu
// set default amount of space for gametype, create game

// play with black, computer is white
void BoardView::newBlackSingleGame()
{
  disconnectAllPlayers();
  space = 1;

  AIPlayer* ai = new AIPlayer(WHITE);
  connectPlayer(ai);
  connect( this, SIGNAL(moveMade(unsigned int)), ai, SLOT(extMove(unsigned int)));

  newGame(BLACK);
}

// play with white, computer is black
void BoardView::newWhiteSingleGame()
{
  disconnectAllPlayers();
  space = 1;

  AIPlayer* ai = new AIPlayer(BLACK);
  connectPlayer(ai);
  connect( this, SIGNAL(moveMade(unsigned int)), ai, SLOT(extMove(unsigned int)));

  newGame(WHITE);
}

// play a two player game with regular rules
void BoardView::newTwoGame()
{
  disconnectAllPlayers();
  space = 0;

  newGame(BOTH);
}

// play a two player game with flip adjacent opposites until blank or own -rules
void BoardView::newWeirdTwoGame()
{
  disconnectAllPlayers();
  space = 0;

  newGame(BOTH, TRUE);
}

// computer vs computer
void BoardView::newDemoGame()
{
  disconnectAllPlayers();
  space = 2;

  AIPlayer* ai = new AIPlayer(WHITE);
  connectPlayer(ai);
  connect( this, SIGNAL(moveMade(unsigned int)), ai, SLOT(extMove(unsigned int)));
  ai = new AIPlayer(BLACK);
  connectPlayer(ai);
  connect( this, SIGNAL(moveMade(unsigned int)), ai, SLOT(extMove(unsigned int)));

  newGame(BLANK);
}

// disconnects player
void BoardView::disconnectPlayer(QObject* exPlayer) {
       disconnect (exPlayer, SIGNAL( disconnectMe(QObject*) ),
                       this, SLOT( disconnectPlayer(QObject*) ));
       disconnect (exPlayer, SIGNAL( makeMove(unsigned int) ),
                       this, SLOT( extMove(unsigned int) ));
       disconnect (this, SIGNAL(moveMade(unsigned int)), exPlayer, SLOT(extMove(unsigned int)));
}



////////
// PRIVATE METHODS
//

// set player (one who clicks on the board), clear board, update view
// player values: BLACK, WHITE, BOTH (two player game), BLANK (ai vs ai)
void BoardView::newGame(unsigned int new_player, bool weird)
{
  player = new_player;
  
  // ditch old rules, make new ones
  std::auto_ptr<GameRules>temp = rules;
  rules = std::auto_ptr<GameRules>( new GameRules(weird) );

  // clear board and update
  clearBoard();
  newMove();
  
  // tell ai to start
  if (player != BLACK) { emit moveMade(0); }
}


// disconnects all player objects
void BoardView::disconnectAllPlayers () {
  QValueList<QObject*>::iterator it;
  for (it = players->begin(); it != players->end(); ++it) {
      disconnectPlayer((*it));
      delete (*it);
  }
  // clear player list
  players->clear();
}


// make move to position pos, tell players move was made
bool BoardView::makeMove(unsigned int pos)
{
  bool retval = rules->setPiece(pos);
  if (retval) {
      newMove();
      emit moveMade( pos ); // tells about the move
  }
  return retval;
}


// do whatever needs to be done when a new move is made
void BoardView::newMove()
{
  updateView();

  // build statusbar signal message
  QString msg;
  if (rules->getTurn() != BLANK) {
    if (rules->getTurn() == BLACK) {
      msg = "Black's";
    } else if (rules->getTurn() == WHITE) {
      msg = "White's";
    }
    if (rules->getTurn() == player) {
      msg += " (your)";
    }
    msg += " turn.";
  } else {
    unsigned int b_score, w_score;
    b_score = rules->getScore(BLACK);
    w_score = rules->getScore(WHITE);

    msg = "Game over! ";

    if (b_score > w_score) {
      msg += "Black wins";
    } else if (b_score < w_score) {
      msg += "White wins";
    } else {
      msg += "Tie";
    }

    msg += " (";
    msg += QString().setNum(b_score);
    msg += " - ";
    msg += QString().setNum(w_score);
    msg += ").";

  }
  emit moveUpdate( msg );
}

File: html/src/app/boardview.ts
import {
  BLANK,
  BLACK,
  WHITE,
  COLUMNS,
  ROWS,
  DEFAULT_BG,
  DEFAULT_BLANK,
  DEFAULT_BLACK,
  DEFAULT_WHITE,
  DEFAULT_BTW,
  DEFAULT_WTB,
  BTW,
  WTB,
} from './constants';
import type { ImageMap } from './types';
import { GameRules } from './gamerules';
import { Piece } from './piece';
import { AnimPiece } from './animpiece';
import { AIPlayer } from './aiplayer';

const BOTH = 999999;

export class BoardView {
  private canvas: HTMLCanvasElement;
  private ctx: CanvasRenderingContext2D;
  private images: ImageMap;
  private rules = new GameRules();
  private pieces: Piece[] = [];
  private wtbPieces: AnimPiece[] = [];
  private btwPieces: AnimPiece[] = [];
  private players: AIPlayer[] = [];
  private animations = true;
  private animSpeed = 1;
  private animRandomJitter = false;
  private animType = 2;
  private player = BLACK;
  private space = 1;
  private statusCallback: (message: string) => void;
  private animationHandle: number | null = null;

  constructor(canvas: HTMLCanvasElement, images: ImageMap, statusCallback: (message: string) => void) {
    this.canvas = canvas;
    const ctx = canvas.getContext('2d');
    if (!ctx) {
      throw new Error('Unable to get 2d context');
    }
    this.ctx = ctx;
    this.images = images;
    this.statusCallback = statusCallback;

    this.canvas.width = 450;
    this.canvas.height = 450;

    for (let i = 0; i < COLUMNS * ROWS; i++) {
      const piece = new Piece();
      piece.setPosition(i);
      piece.setColor(BLANK);
      this.pieces.push(piece);
    }

    for (let i = 0; i < COLUMNS * ROWS; i++) {
      const wtb = new AnimPiece(WTB);
      wtb.setPosition(i);
      this.wtbPieces.push(wtb);
      const btw = new AnimPiece(BTW);
      btw.setPosition(i);
      this.btwPieces.push(btw);
    }

    this.newMove();
    this.newBlackSingleGame();

    this.canvas.addEventListener('click', (event) => this.handleClick(event));
    this.startRenderLoop();
  }

  setAnimated(animate: boolean): void {
    this.animations = animate;
  }

  getAnimated(): boolean {
    return this.animations;
  }

  setAnimSpeed(speed: number): void {
    if (speed >= 0) {
      this.animSpeed = speed;
      this.animRandomJitter = false;
    } else {
      this.animSpeed = 1;
      this.animRandomJitter = true;
    }
  }

  getAnimSpeed(): number {
    return this.animSpeed;
  }

  newBlackSingleGame(): void {
    this.disconnectAllPlayers();
    this.space = 1;

    const ai = new AIPlayer(WHITE);
    this.connectPlayer(ai);
    ai.onMakeMove = (pos) => this.extMove(pos);

    this.newGame(BLACK);
  }

  newWhiteSingleGame(): void {
    this.disconnectAllPlayers();
    this.space = 1;

    const ai = new AIPlayer(BLACK);
    this.connectPlayer(ai);
    ai.onMakeMove = (pos) => this.extMove(pos);

    this.newGame(WHITE);
  }

  newTwoGame(): void {
    this.disconnectAllPlayers();
    this.space = 0;

    this.newGame(BOTH);
  }

  newWeirdTwoGame(): void {
    this.disconnectAllPlayers();
    this.space = 0;

    this.newGame(BOTH, true);
  }

  newDemoGame(): void {
    this.disconnectAllPlayers();
    this.space = 2;

    const aiWhite = new AIPlayer(WHITE);
    this.connectPlayer(aiWhite);
    aiWhite.onMakeMove = (pos) => this.extMove(pos);

    const aiBlack = new AIPlayer(BLACK);
    this.connectPlayer(aiBlack);
    aiBlack.onMakeMove = (pos) => this.extMove(pos);

    this.newGame(BLANK);
  }

  toggleAnim(): void {
    this.setAnimated(!this.animations);
  }

  private updateView(): void {
    if (this.animations) {
      this.animateBoard();
    } else {
      this.clearBoard();
    }
  }

  private animateBoard(): void {
    let i = -1;
    let delay = 0;
    for (const piece of this.pieces) {
      const newFrame = this.rules.getPiece(piece.getPosition());
      if (piece.getColor() !== newFrame) {
        if (newFrame === BLANK || piece.getColor() === BLANK) {
          piece.setColor(newFrame);
          piece.setFrame(newFrame);
        } else {
          piece.setColor(newFrame);
          piece.setFrame(newFrame);
          
          for (const anim of [...this.btwPieces, ...this.wtbPieces]) {
            if (anim.getResult() === newFrame && anim.position === piece.getPosition()) {
              if (!this.animRandomJitter) {
                i = this.animSpeed * this.animDelayAlgo(this.animType, piece, (value) => (delay = value), delay);
              } else {
                i++;
              }
              anim.animPlay(piece, i, this.animRandomJitter);
              break;
            }
          }
          
        }
      }
    }
  }

  private animDelayAlgo(algo: number, piece: Piece, setDelay: (value: number) => void, delay: number): number {
    switch (algo) {
      case 1:
        return 5.5 * Math.log10(3.5 * this.distanceFromNew(piece.getPosition()) - 3.0);
      case 2:
        return this.distanceFromNew(piece.getPosition()) - 1.0;
      case 3:
        return (4 * this.getMaximumDistance()) / (5.5 * Math.log10(3.5 * this.distanceFromNew(piece.getPosition()) - 3.0));
      case 4:
        return (3 * this.getMaximumDistance()) / (1 + Math.abs(this.distanceFromNew(piece.getPosition()))) - 1.0;
      case 5:
        return this.distanceFromNew(piece.getPosition()) - 1.0 + Math.PI + this.angleFromNew(piece.getPosition());
      default:
        const retval = delay;
        setDelay(delay + 1);
        return retval;
    }
  }

  private angleFromNew(pos: number): number {
    const changes = this.rules.getChanges();
    const newpos = changes[1];
    const p = this.rules.uiToTCoord(pos);
    const n = this.rules.uiToTCoord(newpos);

    const diffX = p.x - n.y;
    const diffY = p.y - n.y;

    return Math.atan2(diffX, diffY);
  }

  private getMaximumDistance(): number {
    let maxDistance = 0;
    const changes = this.rules.getChanges();
    for (let idx = 2; idx < changes.length; idx++) {
      const temp = this.distanceFromNew(changes[idx]);
      if (temp > maxDistance) {
        maxDistance = temp;
      }
    }
    return maxDistance;
  }

  private distanceFromNew(pos: number): number {
    const changes = this.rules.getChanges();
    const newpos = changes[1];
    const p = this.rules.uiToTCoord(pos);
    const n = this.rules.uiToTCoord(newpos);

    const absX = Math.abs(p.x - n.x);
    const absY = Math.abs(p.y - n.y);
    return absX < absY ? absY : absX;
  }

  private clearBoard(): void {
    for (const piece of this.pieces) {
      const newFrame = this.rules.getPiece(piece.getPosition());
      piece.setColor(newFrame);
      piece.setFrame(newFrame);
    }
    for (const anim of [...this.btwPieces, ...this.wtbPieces]) {
      anim.stopPlaying();
    }
  }

  private isSpace(): boolean {
    return this.space > 0;
  }

  private connectPlayer(player: AIPlayer): boolean {
    if (this.isSpace()) {
      this.space--;
      this.players.push(player);
      player.onDisconnect = () => this.disconnectPlayer(player);
      return true;
    }
    return false;
  }

  private handleClick(event: MouseEvent): void {
    const rect = this.canvas.getBoundingClientRect();

    // If canvas is CSS-scaled (PWA/mobile), map pointer coords back to internal canvas coords.
    const scaleX = this.canvas.width / rect.width;
    const scaleY = this.canvas.height / rect.height;

    const x = (event.clientX - rect.left) * scaleX;
    const y = (event.clientY - rect.top) * scaleY;

    const col = Math.floor(x / 56);
    const row = Math.floor(y / 56);
    const pos = row * 8 + col;

    const piece = this.pieces[pos];
    if (!piece) {
      return;
    }
    if (piece.getFrame() === BLANK) {
      if (this.player !== BLANK && (this.player === BOTH || this.rules.getTurn() === this.player) && this.makeMove(pos)) {
        // handled
      }
    }
  }

  private extMove(pos: number): void {
    if (this.rules.getTurn() !== this.player || this.player === BLANK) {
      this.makeMove(pos);
    }
  }

  private newGame(newPlayer: number, weird = false): void {
    this.player = newPlayer;
    this.rules = new GameRules(weird);
    this.clearBoard();
    this.newMove();
    if (this.player !== BLACK) {
      this.signalMoveMade(0);
    }
  }

  private disconnectAllPlayers(): void {
    for (const player of this.players) {
      this.disconnectPlayer(player);
    }
    this.players = [];
  }

  private disconnectPlayer(player: AIPlayer): void {
    player.onMakeMove = null;
    player.onDisconnect = null;
  }

  private makeMove(pos: number): boolean {
    const retval = this.rules.setPiece(pos);
    if (retval) {
      this.newMove();
      this.signalMoveMade(pos);
    }
    return retval;
  }

  private signalMoveMade(pos: number): void {
    for (const player of this.players) {
      player.extMove(pos);
    }
  }

  private newMove(): void {
    this.updateView();

    let msg = '';
    if (this.rules.getTurn() !== BLANK) {
      if (this.rules.getTurn() === BLACK) {
        msg = "Black's";
      } else if (this.rules.getTurn() === WHITE) {
        msg = "White's";
      }
      if (this.rules.getTurn() === this.player) {
        msg += ' (your)';
      }
      msg += ' turn.';
    } else {
      const bScore = this.rules.getScore(BLACK);
      const wScore = this.rules.getScore(WHITE);
      msg = 'Game over! ';
      if (bScore > wScore) {
        msg += 'Black wins';
      } else if (bScore < wScore) {
        msg += 'White wins';
      } else {
        msg += 'Tie';
      }
      msg += ` (${bScore} - ${wScore}).`;
    }
    this.statusCallback(msg);
  }

  private startRenderLoop(): void {
    const render = (now: number) => {
      this.render(now);
      this.animationHandle = requestAnimationFrame(render);
    };
    this.animationHandle = requestAnimationFrame(render);
  }

  private render(now: number): void {
    const ctx = this.ctx;
    const bg = this.images[DEFAULT_BG];
    if (bg) {
      ctx.clearRect(0, 0, this.canvas.width, this.canvas.height);
      ctx.drawImage(bg, 0, 0, this.canvas.width, this.canvas.height);
    }

    for (const piece of this.pieces) {
      let sprite = this.images[DEFAULT_BLANK];
      if (piece.getFrame() === BLACK) {
        sprite = this.images[DEFAULT_BLACK];
      } else if (piece.getFrame() === WHITE) {
        sprite = this.images[DEFAULT_WHITE];
      }
      const { x, y } = piece.getPixel();
      ctx.drawImage(sprite, x, y, sprite.width, sprite.height);
    }

    for (const anim of [...this.btwPieces, ...this.wtbPieces]) {
      if (anim.playing) {
        const updated = anim.advanceFrame(now);
        const frames = anim.direction === BTW ? DEFAULT_BTW : DEFAULT_WTB;
        const frameIndex = !updated && anim.currentFrame === 0 ? frames.length-1 : Math.min(anim.currentFrame, frames.length - 1);
        const frame = this.images[frames[frameIndex]];
        const { x, y } = anim.getPixel();
        ctx.drawImage(frame, x, y, frame.width, frame.height);
      }
    }
  }
}
